File: AOC_2024/Python/Day06/day06.py
from sys import argv

from Utils.utils import timer

with open(argv[1] if len(argv) > 1 else "data.txt") as file:
    data = file.read().split()

start = (0, 0)
for row_idx, row in enumerate(data):
    for col_idx, col in enumerate(row):
        if col == '^':
            start = (row_idx, col_idx)


def part1():
    seen = set()
    direction = (-1, 0)
    cur_pos = start
    while True:
        x, y = cur_pos
        dx, dy = x + direction[0], y + direction[1]
        if dx < 0 or dx >= len(data) or dy < 0 or dy >= len(data[0]):
            seen.add(cur_pos)
            break
        while data[dx][dy] == '#':
            direction = (direction[1], -direction[0])
            dx, dy = x + direction[0], y + direction[1]
        seen.add(cur_pos)
        cur_pos = (dx, dy)
    return len(seen)


def find_loop():
    seen = set()
    direction = (-1, 0)
    cur_pos = start
    while True:
        x, y = cur_pos
        dx, dy = x + direction[0], y + direction[1]
        if dx < 0 or dx >= len(data) or dy < 0 or dy >= len(data[0]):
            return False
        while data[dx][dy] == '#':
            direction = (direction[1], -direction[0])
            dx, dy = x + direction[0], y + direction[1]
            if dx < 0 or dx >= len(data) or dy < 0 or dy >= len(data[0]):
                return False
        if ((dx, dy), direction) in seen:
            return True
        seen.add(((dx, dy), direction))
        cur_pos = (dx, dy)


def part2():
    count = 0

    for i, row in enumerate(data):
        row_copy = row
        for j, col in enumerate(row):
            if col != "^" and col != '#':
                data[i] = row[:j] + '#' + row[j + 1:]
                if find_loop():
                    count += 1
            data[i] = row_copy
    return count


if __name__ == '__main__':
    timer(part1, part2)

File: AOC_2024/C++/Day06/day06.cpp
#include <fstream>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

#include "utils.h"
using namespace std;

vector<vector<char> > map;
pair<int, int> start;
set<pair<int, int> > initial_seen;


bool is_loop()
{
    set<pair<pair<int, int>, pair<int, int> > > seen;
    pair dir = make_pair(-1, 0);
    pair cur_pos = start;

    while (true)
    {
        int x = cur_pos.first, y = cur_pos.second;
        int dx = cur_pos.first + dir.first, dy = cur_pos.second + dir.second;
        if (dx < 0 or dx >= map.size() or dy < 0 or dy >= map[0].size())
        {
            return false;
        }
        while (map[dx][dy] == '#')
        {
            const int temp = dir.first;
            dir.first = dir.second;
            dir.second = -temp;
            dx = x + dir.first, dy = y + dir.second;
            if (dx < 0 or dx >= map.size() or dy < 0 or dy >= map[0].size())
            {
                return false;
            }
        }
        if (seen.contains(make_pair(make_pair(dx, dy), dir)))
        {
            return true;
        }
        seen.insert(make_pair(make_pair(dx, dy), dir));
        cur_pos.first += dir.first;
        cur_pos.second += dir.second;
    }
}

long long part1()
{
    pair dir = {-1, 0};
    pair<int, int> cur_pos = start;
    while (true)
    {
        int x = cur_pos.first, y = cur_pos.second;
        int dx = x + dir.first, dy = y + dir.second;
        if (dx < 0 or dx >= map.size() or dy < 0 or dy >= map[0].size())
        {
            initial_seen.insert(cur_pos);

            break;
        }
        while (map[dx][dy] == '#')
        {
            const int temp = dir.first;
            dir.first = dir.second;
            dir.second = -temp;
            dx = x + dir.first, dy = y + dir.second;
        }
        initial_seen.insert(cur_pos);
        cur_pos.first += dir.first;
        cur_pos.second += dir.second;
    }
    return static_cast<long long>(initial_seen.size());
}

long long part2()
{
    long long total = 0;
    for (auto &row: map)
    {
        for (char &ch: row)
        {
            if (ch != '^' and ch != '#')
            {
                char temp = ch;
                ch = '#';
                total += is_loop();
                ch = temp;
            }
        }
    }
    return total;
}

int main()
{
    string line;
    ifstream file("../AOC_2024/C++/Day03/data.txt");

    while (getline(file, line))
    {
        stringstream ss(line);
        vector<char> row;
        char ch;
        while (ss >> ch)
        {
            row.push_back(ch);
        }
        map.push_back(row);
    }

    for (int i = 0; i < map.size(); i++)
    {
        for (int j = 0; j < map[i].size(); j++)
        {
            if (map[i][j] == '^')
            {
                start = make_pair(i, j);
            }
        }
    }

    utils::timer(part1, part2);
}
